### PR TITLE
Check for needless raw strings in `format_args!()` template as well

### LINKED ISSUE #13503: `needless_raw_strings`, `needless_raw_string_hashes`: False negative with macros
### Summary

[`needless_raw_strings`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_strings) and [`needless_raw_string_hashes`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_string_hashes) does not work when the string is inside macros.

### Lint Name

needless_raw_strings, needless_raw_string_hashes

### Reproducer

```rust
#![warn(clippy::needless_raw_strings, clippy::needless_raw_string_hashes)]

let _ = r"SELECT * FROM posts"; // needless-raw-strings
let _ = r#"SELECT * FROM posts"#; // needless-raw-strings
let _ = r##"SELECT * FROM "posts""##; // needless-raw-string-hashes

// None of lines below triggers any warning
println!(r"SELECT * FROM posts");
println!(r#"SELECT * FROM posts"#);
println!(r##"SELECT * FROM "posts""##);

// No warning when used with sqlx
sqlx::query!(
    r#"
    SELECT * FROM posts
    "#,
)
.fetch_all(&pool)
.await
.unwrap();
```

### Version

rustc 1.83.0-nightly (363ae4188 2024-09-24)
binary: rustc
commit-hash: 363ae4188316b8b22cf6c1890bc73d84d05f70a4
commit-date: 2024-09-24
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0

### PR DESCRIPTION
**Pull request by samueltardieu**
changelog: [`needless_raw_strings`, `needless_raw_string_hashes`]: check `format_args!()` template as well

Fix #13503

### CODE REVIEW

**Comment by samueltardieu** [comment]
It looks like lintcheck detects an ICE in the _base_ as well, which would mean it is not due to this PR, right? Interestingly, I don't seem to be able to reproduce it locally.

**Comment by Alexendoo** [comment]
Looks like futures-util 0.3.30 doesn't compile on our nightly

```
error[E0310]: the parameter type `Fut` may not live long enough
  --> src/stream/futures_unordered/task.rs:81:9
   |
81 |         waker_ref(this)
   |         ^^^^^^^^^^^^^^^
   |         |
   |         the parameter type `Fut` must be valid for the static lifetime...
   |         ...so that the type `Fut` will meet its required lifetime bounds
   |
help: consider adding an explicit lifetime bound
   |
78 | impl<Fut: 'static> Task<Fut> {
   |         +++++++++
```

You could bump it to 0.3.31, it [looks](https://github.com/rust-lang/futures-rs/pull/2830) fixed there

**Comment by samueltardieu** [comment]
> Looks like futures-util 0.3.30 doesn't compile on our nightly

I noticed this but was surprised to see it identified as an ICE. I will bump it.

**Comment by samueltardieu** [comment]
Done, and also submitted separately in #13505

File: clippy_lints/src/raw_strings.rs [hunk 1]
```diff
@@ -71,6 +71,23 @@ impl RawStrings {
 
 impl EarlyLintPass for RawStrings {
     fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::FormatArgs(format_args) = &expr.kind
+            && !in_external_macro(cx.sess(), format_args.span)
+            && format_args.span.check_source_text(cx, |src| src.starts_with('r'))
+            && let Some(str) = snippet_opt(cx.sess(), format_args.span)
+        {
+            let count_hash = str.bytes().skip(1).take_while(|b| *b == b'#').count();
+            let str = &str[count_hash + 2..str.len() - count_hash - 1];
```
Comment (Alexendoo) [inline clippy_lints/src/raw_strings.rs:80]: Let's go with a `.get` to guard against any proc macro shenanigans causing a panic

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Thanks!

@bors r+

**Pull request merged**